### ISSUE TITLE
Hunting for memory leak

### ISSUE THREAD

**Issue by vasily-kirichenko** [Hopac/Hopac#107]
I've managed to make a minimal repro. So, we have a couple of servers which waits for messages from same couple of channels, but in difference precedence: first server (`worker`) prefer the first channel, second one - the second channel.

``` fsharp
open Hopac
open Hopac.Infixes
open System

type State = { Count: int }

[<EntryPoint>]
let main _ = 
    let worker alt =
        Job.iterateServer { Count = 0 } (fun s -> 
            alt ^-> fun _ -> { s with Count = s.Count + 1 }
        ) |> queue

    let ch1, ch2 = Ch(), Ch()

    let _worker1 = worker <| Alt.choose [ch1; ch2]
    let _worker2 = worker <| Alt.choose [ch2; ch1]

    while true do
        printfn "Press a key to send a task"
        Console.ReadKey() |> ignore
        ch1 *<- () |> run
    0
```

After 10 iterations dotMemory shows the following picture:
![image](https://cloud.githubusercontent.com/assets/873919/16176136/726f22de-360e-11e6-9290-8cbdaf9bb27d.png)

Retention paths for `State` objects, for example, are as follows:
![image](https://cloud.githubusercontent.com/assets/873919/16176152/004fdc92-360f-11e6-9b5c-dd91eabafc39.png)
![image](https://cloud.githubusercontent.com/assets/873919/16176155/0c0bb056-360f-11e6-9fb1-39be087c1ec7.png)

If I comment out `let _worker2 = worker <| Alt.choose [ch2; ch1]`, leaving only one worker, the leak is completely gone.

The real task is as following:
- there are a number of channels
- there are a number of workers, each of which should always prefer messages from one of the channels, if there is no messages in that channel, the worker should get messages from all other channels, randomly
  In the real code workers synchronize on things like this: `preferableCh <|> Alt.chooser otherChs`, but the minimal repro above demonstrate the leak as well.

**Comment by vasily-kirichenko** [comment]
@polytypic I've managed to make a minimal repro, could you please take a look at this?

**Comment by vasily-kirichenko** [comment]
Addition: sending a message to `ch2` like `ch2 *<- () |> run` frees up all the stuck objects, but in the real code there are channels which are not given messages for days, so it's not a solution.

What's interesting, adding `timeOutMillis 1000 ^->. s` alternative to the worker does not solve the problem.

**Comment by polytypic** [comment]
> The real task is as following:
>
> * there are a number of channels
>
> * there are a number of workers, each of which should always prefer messages
>   from one of the channels, if there is no messages in that channel, the worker
>   should get messages from all other channels, randomly In the real code workers
>   synchronize on things like this: preferableCh <|> Alt.chooser otherChs, but
>   the minimal repro above demonstrate the leak as well.

So, the alternative each worker is using looks like this:

```fs
myPreferredChannel <|> Alt.chooser otherChannels
```

The essential feature above is the use of randomized choice over a collection of channels.  It probably does not work the way one might assume it to work.

Randomization only matters in case the alternatives are already available at the time the choice is instantiated.  In case of channels, if you have a collection of channels that are all "empty", i.e. they have no available givers, then randomization has no effect.  In case of channels, randomization over a collection of channels only matters in case some channels are non-empty.

Also, channels always work in a FIFO manner for both givers and takers.

What basically happens with the above pattern is that each worker first "checks" their dedicated channel.  They get a message from their dedicated channel only when both of the following conditions hold:

1. no other worker is queued to take a message from the channel, and
2. there is a giver on the channel.

The above two conditions are unlikely to hold most of the time.  The conditions basically hold most of the time only when the channels are all non-empty most of the time.

In case the conditions do not hold, such as when the channels are empty most of the time, then instead of the preferred + random choice, you actually get FIFO behavior.  The workers are queued to take messages from the channels and the probability of a worker getting a message from its preferred channel, or any other channel, approaches `1/n` where `n` is the number of workers.  Even when a worker dedicated to a particular channel is available, it will have to wait in queue until all the other workers queued earlier to that channel have gotten messages.

I assume that the idea has been that whenever a worker is available (i.e. not busy) then that worker will take messages given on its dedicated channel.  (The above pattern does not achieve this.)  You can achieve this property using a different arrangement.  For example, by using a dedicated channel for each worker and a single shared channel for all workers rather than a choice over channels.  To send a message with a preference to a particular worker, one makes a choice of either first giving the message on the dedicated channel of the desired worker or on the channel shared by all workers.

This pattern, using a shared channel, both works and is also faster.  With the choice over channels, each worker performs `O(n)` work, where `n` is the number of workers, to queue waiting for messages on the channels.  With the shared channel, each worker performs only `O(1)` work (queues itself on two channels). In both cases sending a messages is `O(1)`.

Here is a script that demonstrates both approaches:

```fs
#I "packages/Hopac/lib/net45"
#I "packages/FSharp.Core/lib/net40"

#r "Hopac.Core.dll"
#r "Hopac.Platform.dll"
#r "Hopac.dll"

open System
open System.Threading
open Hopac
open Hopac.Infixes

let inline (^) x = x

let sync = obj ()
let inline printfns fmt =
  fmt
  |> Printf.kprintf ^ fun s ->
       lock sync ^ fun () ->
         Console.WriteLine s

let worker name alt =
  server ^ Job.iterate 0 ^ fun count ->
  printfns "    (%s available)" name
  alt ^-> fun i ->
    printfns "    %s at %d got %d" name count i
    count + i

let runs (sleep: option<int>) op =
  run op
  sleep
  |> Option.iter Thread.Sleep

let nWorkers = 10
let nMsgs = 15

let QueueLike =
  let chs = Array.init nWorkers (ignore >> Ch)
  do for i=0 to nWorkers-1 do
       let anyOtherCh =
         Array.append
            <| Array.sub chs 0     i
            <| Array.sub chs (i+1) (nWorkers-(i+1))
         |> Alt.chooser
       let myCh = chs.[i]
       myCh <|> anyOtherCh |> worker ^ sprintf "Q%d" i
  fun i m ->
    chs.[i] *<- m

let Preferred =
  let sharedCh = Ch ()
  let chs = Array.init nWorkers (ignore >> Ch)
  do for i=0 to nWorkers-1 do
       let myCh = chs.[i]
       myCh <|> sharedCh |> worker ^ sprintf "P%d" i
  fun i m ->
    chs.[i] *<- m <|> sharedCh *<- m

let demo name op =
  printfns "%s no delay" name
  for i=1 to nMsgs do
    printfns "  Sending %d" i
    op 0 i |> runs None
  Thread.Sleep 1000
  printfns "%s with delay" name
  for i=1 to nMsgs do
    printfns "  Sending %d" i
    op 0 i |> runs ^ Some 10

do demo "QueueLike" QueueLike
do demo "Preferred" Preferred
do Thread.Sleep 1000
   printfns "Done"
```

BTW, this approach would not work with Concurrent ML.

**Comment by vasily-kirichenko** [comment]
Thanks for the detailed answer! 

I cannot test the shared channel approach right now, but I think it will fix the memory leak. 

However, the real task (and code) is a bit more complicated, it was a mistake to not show you the full task.

The application consumes tasks from outside (RabbitMQ), each task has a unique ID and a Slot ID (SID). SID is needed for partitioning the app resources to a number of "slots", each of which is nothing else but a worker pool of fixed size. So, for example, we have 3 pools: 

* SID = -1, Capacity = 10
* SID = 0, Capacity = 1
* SID = 1, Capacity = 4

Slots -1 and 1 can "steal" tasks from other slots, slot 0 - cannot, because it must always be ready to process tasks with SID = 0 with no delay.

Each task, as a result of its processing, produces zero to possibly hundreds of thousands sub-tasks (with same ID), which must be processed the same way as original "root" task. This process is recursive and can be up to ~2000 levels. The work stealing is needed here to avoid the situation when a "fat" task (I mean it will produce many sub tasks) comes to say SID = 0 and, as it has capacity = 1, all the sub tasks will be processed by single worker, which is very slow and a clear waste of resources.

A naive approach could be as following:

* a channel and `capacity` number of workers are created for each slot 
* a shared channel is created
* each worker waits on `myChannel <|> sharedChannel`
* new tasks and sub tasks are "given" by `channelForSID *<- m <|> sharedChannel *<- m` (sub tasks are given using something like `subTasks |> Seq.Con.iterJobIgnore (Ch.give ...)` - `give` here instead of `send` is because we must wait for all sub tasks have been given to one channel or another in order to determine when the whole recursive process is finished). 

OK, it looks good.

Now the problem. All tasks must be processed evenly, I mean workers must take tasks/subtasks with different IDs on random or round robin fashion (I already described it on another issue here). It's needed to avoid the situation when one "fat" task flooded all channels (I mean it created a lot of `Job`s each of which is trying to `give` a sub task to both the dedicated (for its SID) channel and to the shared channel). It results with unwanted behavior: all other tasks (with different IDs) will be processed after all sub tasks of the "fat" task are done.

To solve the problem, a dedicated channel is created for _each task ID_. After that workers are getting messages by `Alt.chooser allActiveTaskIDChs <|> shared`, and messages are given by `channelForThisTaskID *<- m <|> sharedCh *<- m`.

Does it make sense? (I'll test it for correctness and leaks in three or four days).

**Comment by vasily-kirichenko** [comment]
OK, I removed a-channel-for-task-id stuff for now.

**Comment by haf** [comment]
Will close this, because the PR has merged and is being released. @vasily-kirichenko Can you test the new release?
